libraries/chain/include/graphene/chain/hardfork_visitor.hpp

```diff
@@ -0,0 +1,83 @@
+#pragma once
+/*
+ * Copyright (c) 2019 Contributors
+ *
+ * The MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <graphene/protocol/operations.hpp>
+
+#include <graphene/chain/hardfork.hpp>
+
+#include <fc/reflect/typelist.hpp>
+
+#include <type_traits>
+#include <functional>
+
+namespace graphene { namespace chain {
+using namespace protocol;
+namespace TL { using namespace fc::typelist; }
+
+/**
+ * @brief The hardfork_visitor struct checks whether a given operation type has been hardforked in or not
+ *
+ * This visitor can be invoked in several different ways, including operation::visit, typelist::runtime::dispatch, or
+ * direct invocation by calling the visit() method passing an operation variant, narrow operation type, operation tag,
+ * or templating on the narrow operation type
+ */
+struct hardfork_visitor {
```

Comment:
I like how this was done. Elegant. However, `hardfork_visitor` is a fairly generic name for a specific purpose. Perhaps it can be "genericized" for more than BSIP 40? Or a name change? Or put in a namespace?

Reply:
Yeah, this is an annoying type to name... It's a visitor on operation, so maybe `operation_hardfork_visitor` is better, or the wordier `operation_is_hardforked_in_visitor`? xP

Reply:
It's already available to `chain` at large; there's nothing BSIP-40 specific about it.